### PR TITLE
Add ActivityView callback (completion handler)

### PR DESCRIPTION
**Pull request by cmcewen**
Thanks for creating this - wanted to do some tracking on whether a user completed an activity, so I implemented a callback with the results of the ActivityView. It's a bit hacky to make the callback optional, but I didn't want to break the API. If this looks good I'll update the Readme to explain more.

### CODE REVIEW

File: ActivityView.ios.js [hunk 1]
```diff
@@ -4,19 +4,48 @@
...
 var ActivityView = {
-  show: NativeActivityView.show
+   show(shareObject: Object, callback: Function) {
+    invariant(
+      typeof shareObject === 'object' && shareObject !== null,
+      'Must provide shareObject to show ActivityView'
+    );
+    invariant(
+      (typeof callback === 'function' || callback === undefined),
+      'Callback must be a function'
```
Comment (chirag04) [inline ActivityView.ios.js]: if callback is optional then is this invariant required?

File: ActivityView.ios.js [hunk 2]
```diff
@@ -4,19 +4,48 @@
...
+    );
+
+    NativeActivityView.show(
+      shareObject,
+      callback ? callback : function() {}
```
Comment (chirag04) [inline ActivityView.ios.js]: IMO, we should just do this check in objective-c.

File: ActivityView.m [hunk 3]
```diff
@@ -54,6 +55,15 @@ - (dispatch_queue_t)methodQueue
...
     UIActivityViewController *activityView = [[UIActivityViewController alloc] initWithActivityItems:shareObject applicationActivities:nil];
+
+    [activityView setCompletionWithItemsHandler:
+        ^(NSString *activityType, BOOL completed, NSArray *returnedItems, NSError *activityError) {
+            callback(@[RCTNullIfNil(activityType),
+                       @(completed),
+                       RCTNullIfNil(returnedItems),
+                       RCTNullIfNil(activityError.localizedDescription)]);
+        }
+    ];
```
Comment (chirag04) [inline ActivityView.m]: register the completionHandler only if you have callback.

File: ActivityView.ios.js [hunk 4]
```diff
@@ -4,19 +4,48 @@
  */
 'use strict';
 
-var React = require('react-native');
-var {
-  NativeModules
-} = React;
+var invariant = require('invariant');
```
Comment (chirag04) [inline ActivityView.ios.js]: we can really get rid of this. propType can help if it's really important.

File: ActivityView.ios.js (hunk 2 above)
Comment (grabbou) [inline ActivityView.ios.js]: Then it means no flow support for people using it ;)

File: ActivityView.ios.js (hunk 4 above)
Comment (naoufal) [inline ActivityView.ios.js]: Agreed.

**Comment by naoufal** [comment]
@cmcewen Thanks for your contribution.  Can you confirm that this resolves #14?

**Comment by cmcewen** [comment]
Hi @chirag04 and @naoufal - apologies for taking so long to follow up on this. I reworked the PR so that it adds a new function, mainly due to the fact that there is no way to include an optional callback nicely. This does resolve #14, and also includes the activityType selected by the user in the callback. It also removes the invariant requirement based on your comments.

**Comment by naoufal** [comment]
@cmcewen I've been thinking about this and I think I might prefer using the EventEmitter for this.  Something like [this](https://github.com/naoufal/react-native-safari-view#events).

Reason is, I'd like to avoid have two methods that have similar functionality. React Native also uses something similar for Keyboard Events.

Would this implementation address the use case you were addressing with this PR?

**Comment by cmcewen** [comment]
@naoufal it would be possible to use the event emitter, but I'm not sure that is an appropriate API for this use case. 

iOS natively has both completion handlers and the NSNotificationCenter, which I think correspond to callbacks and events. In this case the activity view uses a completion handler, and I think it would make more sense to use a callback on the JS side as well.

The activity view brings up a separate UI, and there is no possible interaction until either an action is taken or it is cancelled, so I feel like it makes sense to handle that interaction as a callback.

On the other hand, something like a keyboard typing or a movie playing (which I know natively requires subscribing to notifications) seems like a better fit to use events.

I agree that having two separate methods that act similarly is not ideal. In that case, I suggest either the callback is a required parameter, or an empty callback is automatically included (which is what the first version of this PR used)

**Comment by naoufal** [comment]
Fair enough. 

In that case, let's pass the callback in the options object of the `show` method.

**Comment by ilanasufrin** [comment]
Is this still being actively worked on?

**Comment by cmcewen** [comment]
@ilanasufrin I've just been using my fork - at this point the functionality is working fine, just figuring out the right API

@naoufal passing in a function in options still feels a bit weird to me - also i'm not sure exactly how RCTConvert works with functions not passed in with the function call. How about returning a promise? AsyncStorage uses Promises (https://github.com/facebook/react-native/blob/master/Libraries/Storage/AsyncStorage.js#L44) and it seems like other pieces of core are being refactored into promises as well https://github.com/facebook/react-native/pull/5512

**Comment by grabbou** [comment]
I think we could just rewrite `show` function to be `show(opts, callback)` and that's all.

**Comment by cmcewen** [comment]
@grabbou that was actually the very first version of this PR I submitted https://github.com/naoufal/react-native-activity-view/pull/21#discussion_r40797323

**Comment by tomprogers** [comment]
Is there any interest in resolving the conflicts and merging this down? I still have a need for this functionality.

**Pull request closed without merging**